src/request_ext.rs

```diff
@@ -0,0 +1,295 @@
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use ureq::{http, Agent, RequestExt, Error};
+    /// use std::time::Duration;
+    /// let agent = Agent::config_builder()
+    ///     .timeout_global(Some(Duration::from_secs(30)))
+    ///     .build()
+    ///     .new_agent();
+    ///
+    /// let request: Result<http::Response<_>, Error> = http::Request::builder()
+    ///             .method(http::Method::GET)
+    ///             .uri("http://foo.bar")
+    ///             .body(())
+    ///             .unwrap()
+    ///             .with_agent(&agent)
+    ///             .run();
+    /// ```
+    /// # Example with further customizations
+    ///
+    /// In this example we use a specific agent, but apply a request-specific configuration on top.
+    ///
+    /// ```
+    /// use ureq::{http, Agent, RequestExt, Error};
+    /// use std::time::Duration;
+    /// let mut agent = Agent::config_builder()
+    ///     .timeout_global(Some(Duration::from_secs(30)))
+    ///     .build()
+    ///     .new_agent();
+    ///
+    /// let request: Result<http::Response<_>, Error> = http::Request::builder()
+    ///             .method(http::Method::GET)
+    ///             .uri("http://foo.bar")
+    ///             .body(())
+    ///             .unwrap()
+    ///             .with_agent(&mut agent)
+    ///             .configure()
+    ///             .http_status_as_error(false)
+    ///             .run();
+    /// ```
+    fn with_agent<'a>(self, agent: impl Into<AgentRef<'a>>) -> WithAgent<'a, S>;
+}
+
+/// Wrapper struct that holds a [`Request`] associated with an [`Agent`].
+pub struct WithAgent<'a, S: AsSendBody> {
+    pub(crate) agent: AgentRef<'a>,
+    pub(crate) request: Request<S>,
+}
+
+impl<'a, S: AsSendBody> WithAgent<'a, S> {
+    /// Returns a [`ConfigBuilder`] for configuring the request.
+    ///
+    /// This allows setting additional request-specific options before sending the request.
+    pub fn configure(self) -> ConfigBuilder<RequestExtScope<'a, S>> {
+        ConfigBuilder(RequestExtScope(self))
+    }
+
+    /// Executes the request using the associated [`Agent`].
+    pub fn run(self) -> Result<Response<Body>, Error> {
+        self.agent.run(self.request)
+    }
+}
+
+impl<'a, S: AsSendBody> WithAgent<'a, S> {
+    pub(crate) fn request_level_config(&mut self) -> &mut Config {
+        let request_level_config = self.request
+            .extensions_mut()
+            .get_mut::<RequestLevelConfig>();
+
+        if request_level_config.is_none() {
+            self.request.extensions_mut().insert(self.agent.new_request_level_config());
+        }
+
+        // Unwrap is safe because of the above check
+        let req_level: &mut RequestLevelConfig = self.request.extensions_mut().get_mut::<RequestLevelConfig>().unwrap();
+
+        &mut req_level.0
+    }
+}
+
+/// Reference type to hold an owned or borrowed [`Agent`].
+pub enum AgentRef<'a> {
+    Owned(Agent),
+    Borrowed(&'a Agent),
```

Comment:
This borrowed agent was an `&'a mut Agent` in your PR. I don't think there's any situation where we actually mutate the agent, so I figured it was easier to just hold an agent, as it'll also allow the agent to be used across multiple requests in parallel.

Reply:
Sounds good! The pool is behind an `Arc<Mutex<>>`, so it makes sense.